src/generate/src.rs

```diff
@@ -0,0 +1,329 @@
+}
+
+pub trait ToSrc {
+    fn to_src(&self) -> Src;
+}
+
+pub use crate::__generate_src_quotable_to_src as quotable_to_src;
+#[macro_export]
+macro_rules! __generate_src_quotable_to_src {
+    ($ty:ty) => {
+        impl ::proc_quote::ToTokens for $ty {
+            fn to_tokens(&self, tokens: &mut ::proc_macro2::TokenStream) {
+                self.to_src().to_tokens(tokens);
+            }
+            fn into_token_stream(self) -> ::proc_macro2::TokenStream {
+                self.to_src().into_token_stream()
+            }
+        }
+    };
+}
+
+pub use crate::__generate_src_quote as quote;
+#[macro_export]
+macro_rules! __generate_src_quote {
+    ($($t:tt)*) => { $crate::generate::src::Src::new(::proc_quote::quote!($($t)*)) };
+}
+
+// Pretty-printing support.
+
+#[derive(Clone, Debug)]
+enum Elem {
+    Char(char, Spacing),
+    Ident(Ident),
+    Literal(Literal),
+}
+
+impl fmt::Display for Elem {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Elem::Char(c, _) => c.fmt(f),
+            Elem::Ident(i) => i.fmt(f),
+            Elem::Literal(l) => l.fmt(f),
+        }
+    }
+}
+
+impl Elem {
+    fn try_from_tt(tt: TokenTree) -> Result<Elem, (Delimiter, TokenStream)> {
+        Ok(match tt {
+            TokenTree::Group(g) => return Err((g.delimiter(), g.stream())),
+            TokenTree::Ident(i) => Elem::Ident(i),
+            TokenTree::Literal(l) => Elem::Literal(l),
+            TokenTree::Punct(p) => Elem::Char(p.as_char(), p.spacing()),
+        })
+    }
+}
+
+#[derive(Default)]
+struct Line {
+    indent: usize,
+    elems: Vec<Elem>,
+}
+
+impl fmt::Display for Line {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for _ in 0..self.indent {
+            write!(f, "    ")?;
+        }
+        for elem in &self.elems {
+            elem.fmt(f)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+pub struct Fragment {
+    before: Vec<Elem>,
+    lines: Vec<Line>,
+}
+
+impl fmt::Display for Fragment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for elem in &self.before {
+            elem.fmt(f)?;
+        }
+        for line in &self.lines {
+            write!(f, "\n{}", line)?;
+        }
+        Ok(())
+    }
+}
+
+impl Fragment {
+    fn last(&self) -> &[Elem] {
+        self.lines.last().map_or(&self.before, |line| &line.elems)
+    }
+
+    fn last_mut(&mut self) -> &mut Vec<Elem> {
+        self.lines
+            .last_mut()
+            .map_or(&mut self.before, |line| &mut line.elems)
+    }
+
+    fn push(&mut self, elem: Elem) {
+        self.last_mut().push(elem);
+    }
+
+    fn append(&mut self, other: Fragment) {
+        self.last_mut().extend(other.before);
+        self.lines.extend(other.lines);
+    }
+}
+
+impl Src {
+    pub fn pretty_print(&self) -> Fragment {
+        self.pretty_print_in_group(Delimiter::None)
+    }
+
+    fn pretty_print_in_group(&self, outer_delim: Delimiter) -> Fragment {
```

Comment:
This function is quite large. To improve readability, consider splitting up if you can find some sensible units.